exonum/src/runtime/types.rs

```diff
@@ -483,12 +519,39 @@ pub struct InstanceState {
     pub spec: InstanceSpec,
     /// Service instance activity status.
     pub status: InstanceStatus,
+    /// Next status of instance if the value is not `None`.
+    pub next_status: InstanceStatus,
 }
 
 impl InstanceState {
     /// Creates a new instance state with the given specification and status.
     pub fn new(spec: InstanceSpec, status: InstanceStatus) -> Self {
-        Self { spec, status }
+        Self {
+            spec,
+            status,
+            next_status: InstanceStatus::None,
+        }
+    }
+
+    /// Indicates whether the service instance current or pending status is active.
+    pub fn is_active(&self) -> bool {
+        self.status.is_active() || self.next_status.is_active()
+    }
+
+    /// Sets next status as current and changes next status to `None`
+    ///
+    /// # Panics
+    ///
+    /// - If next status is already `None`.
+    pub(crate) fn commit_next_status(&mut self) {
+        assert_ne!(
```

Comment:
To determine `None`-status (if we really need it), I'd suggest to use `Option`. And should we really panic here, isn't it safe to just return an error?